src/main/java/org/folio/auth/authtokenmodule/MainVerticle.java

```diff
@@ -456,316 +432,272 @@ private void handleAuthorize(RoutingContext ctx) {
-      }
-      try {
-        candidateToken = tokenCreator.createJWTToken(dummyPayload.encode());
-      } catch(Exception e) {
-        endText(ctx, 500, "Error creating candidate token: ", e);
-        return;
-      }
-    }
-
     final String authToken = candidateToken;
     logger.debug("Final authToken is {}", authToken);
-    final String errMsg = "Invalid token";
-    try {
-      if (!tokenCache.contains(authToken)) {
-        tokenCreator.checkJWTToken(authToken);
-        tokenCache.add(authToken);
-      }
-    } catch (ParseException p) {
-      logger.error("Malformed token: {}", authToken, p);
-      endText(ctx, 401, errMsg);
-      return;
-    } catch (JOSEException j) {
-      logger.error("Unable to verify token token {}, {}", authToken, j.getMessage());
-      endText(ctx, 401, errMsg);
-      return;
-    } catch (BadSignatureException b) {
-      logger.error("Unsupported JWT format", b);
-      endText(ctx, 401, errMsg);
-      return;
-    }
 
-    JsonObject tokenClaims = getClaims(authToken);
+    Future<Token> tokenValidationResult = Token.validate(authToken, tokenCreator, ctx.request());
 
-    String username = tokenClaims.getString("sub");
-    String jwtTenant = tokenClaims.getString("tenant");
-
-    if (jwtTenant == null || !jwtTenant.equals(tenant)) {
-      logger.error("Expected tenant: {}, got tenant {}", tenant, jwtTenant);
-      endText(ctx, 403, "Invalid token for access");
+    tokenValidationResult.onFailure(h -> {
+      var e = (TokenValidationException)h;
+      logger.error("Invalid token: {}", e.toString());
+      endText(ctx, e.httpResponseCode, "Invalid token");
       return;
-    }
-
-    String tokenUserId = tokenClaims.getString(TOKEN_USER_ID_FIELD);
-    if (tokenUserId != null) {
-      if (userId != null) {
-        if (!userId.equals(tokenUserId)) {
-          endText(ctx, 403,
-           "Payload user id of '" + tokenUserId + "' does not match expected value");
-          return;
+    });
+    
+    tokenValidationResult.onSuccess(h -> {
+      Token token = (Token)h;
+      logger.debug("Validated token of type: {}", token.getClaims().getString("type"));
+
+      String username = token.getClaims().getString("sub");
+
+      // TODO I believe this block solves the issue that Adam found with functional testing.
+      // My approach of just making userId be equal to the X-Okapi-User-Id didn't
+      // give what is here and in the existing code called the finalUserId a value
+      // in this specific case of the tokenUserId having a value _and_ the X-Okapi-User-Id
+      // being null. I'd likek to better understand the circumstances under which these
+      // two variables can have this state and what it means.
+      String temporaryUserId = userId;
+      String tokenUserId = token.getClaims().getString("user_id");
+      if (tokenUserId != null && userId == null) {
+          temporaryUserId = tokenUserId;
+      }
+      final String finalUserId = temporaryUserId;
```

Comment:
```
String headerUserId = ctx.request().headers().get(XOkapiHeaders.USER_ID);
```

```
String tokenUserId = token.getClaims().getString("user_id");
final String userId = headerUserId != null ? headerUserId : tokenUserId;
```

For me this is more readable.

Reply:
Agree the one liner is more readable. Still find it somewhat troubling that we don't have a single source of truth for the userId. Seems fine that that source of truth be in the header _or_ the token, but what bothers me (and led to a difficult bug because I assumed that it would be in the header) was that it wasn't always defined in the header if it was known.